### ISSUE TITLE
Decrypt Spring Cloud Vault Authentication Token 

### ISSUE THREAD

**Issue by bhavanij** [spring-projects/spring-vault#221]
Hello,
We have a requirement for encrypting Vault authentication token credentials used for Vault login.

Here is my sample project: https://github.com/bhavanij/sample-spring-cloud-vault-service

To run this project: 
1.  Add this environment variable
```
ENCRYPT_KEY=1B61EA9E82B5427E990FCAA21C040270
```

2. Insert the following test credentials to Vault (to be read from sample application)
```
vault read secret/value-capture-test/np/db
Key                 Value
---                 -----
refresh_interval    768h
vault.password      TEST_PWD
vault.username      TEST_USER
```
I have couple of questions/issues:

1. To authenticate Vault with decrypted token, I had to write custom `VaultConfig` as shown in the sample application.

How can Spring Vault handle decryption of authentication token without custom VaultConfig?
Please delete `MyVCAbstractVaultConfiguration` in `VaultConfig` in my sample application to reproduce the issue. Should receive `403: Permission Denied` error because Spring is trying to access Vault with cipher token text (without decryption)

2. When debugging the spring code in detail, I figured out that the order in which BootstrapConfigurations are loaded is not as expected. 

BootstrapApplicationListener is loading list of all BootstrapConfiguration class names here:
```
List<String> names = SpringFactoriesLoader
        .loadFactoryNames(BootstrapConfiguration.class, classLoader);
for (String name : StringUtils.commaDelimitedListToStringArray(
        environment.getProperty("spring.cloud.bootstrap.sources", ""))) {
    names.add(name);
}
```
```
org.springframework.cloud.bootstrap.config.PropertySourceBootstrapConfiguration
org.springframework.cloud.bootstrap.encrypt.EncryptionBootstrapConfiguration
org.springframework.cloud.autoconfigure.ConfigurationPropertiesRebinderAutoConfiguration
org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration
org.springframework.cloud.vault.config.DiscoveryClientVaultBootstrapConfiguration
org.springframework.cloud.vault.config.VaultBootstrapConfiguration
org.springframework.cloud.vault.config.VaultBootstrapPropertySourceConfiguration
```
And then `AnnotationAwareOrderComparator.sort(sources);` is sorting those Configuration class names in the highest order first because of which VaultBootstrapConfiguration is executed first and then PropertySourceBootstrapConfiguration and EncryptionBootstrapConfiguration (which I think is incorrect)
```
org.springframework.cloud.vault.config.VaultBootstrapPropertySourceConfiguration
org.springframework.cloud.vault.config.VaultBootstrapConfiguration
org.springframework.cloud.vault.config.DiscoveryClientVaultBootstrapConfiguration
org.springframework.cloud.bootstrap.config.PropertySourceBootstrapConfiguration
org.springframework.cloud.bootstrap.encrypt.EncryptionBootstrapConfiguration
org.springframework.cloud.autoconfigure.ConfigurationPropertiesRebinderAutoConfiguration
org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration
```

3. Though order of `PropertySourceBootstrapConfiguration`  is defined in its class with highest precedence
`private int order = Ordered.HIGHEST_PRECEDENCE + 10;`
after sorting, its precedence is after VaultConfiguration as BootstrapListener is unable to fetch order of `PropertySourceBootstrapConfiguration`. 
Debugging more found an issue with OrderComparator.findOrder() which is returning null for those configurations implementing `Ordered` interface. 
Please check `TestClass` in my sample application to verify this issue
```
	protected Integer findOrder(Object obj) {
		return (obj instanceof Ordered ? ((Ordered) obj).getOrder() : null);
	}
```

4. Any reason for not defining order for `EncryptionBootstrapConfiguration`. If no order defined then that configuration class is given lowest order precedence `because of which VaultConfiguration is executed before `PropertySourceBootstrapConfiguration` and `EncryptionBootstrapConfiguration`. 
Shouldn't `PropertySourceBootstrapConfiguration` and `EncryptionBootstrapConfiguration` be always executed first with highest precedence before any other BootstapConfigurations?

Thank you in advance.

**Comment by mp911de** [comment]
Thanks a lot for the detailed ticket.

## TL;DR

Using encrypted vault tokens does not work and won't work without customization.

## Explanation

To understand why Vault property sources and Encryption work at all, we need to take a look how things are built right now. Properties are bound to `ConfigurationProperties` beans during their bootstrap. At that time, all property sources are already in place. There's very little control over property source post-processing because property sources are installed at a very early time to properly bind configuration properties.

Because property source setup is so tight, Spring Cloud came up with the bootstrap context. All components that want to contribute to infrastructural aspects (such as property sources) are configured in the bootstrap context to then serve for the application context. There are some ways to order configuration classes within the bootstrap context but this affects more which config class can provide its beans first. Once all beans are set up from the bootstrap context, post processors are invoked to do the last preparation steps before the application context gets initialized.

Because of how things are designed, you can't really benefit from encrypted properties as the decryption bit kicks in too late for what you plan to do.

To satisfy your requirement, you could provide your own configuration class that provides a `ClientAuthentication`. Inside the config class, you can parse the actual property value, optionally decrypt it via `TextEncryptor`.

**Comment by bhavanij** [comment]
Thanks a lot for your inputs. I will use custom configuration class to decrypt Vault token to be used in `ClientAuthentication` as you suggested. 

And while debugging Spring code, I found an issue 3 with `OrderComparator.findOrder()` as explained in point 3. 
Please check `TestClass` in my sample application to verify this issue.

**Comment by bhavanij** [comment]
Yes you are right. `obj can't be an instance of Ordered since it's a Class.` But in Spring `BootstrapApplicationListener`, `Class` is mapped to `obj` that is the reason `OrderComparator.findOrder()` is returning `false`. As per my understanding this seems to be a bug in `BootstrapApplicationListener` class.  

Please check `sources` argument for `AnnotationAwareOrderComparator.sort(sources);` code in `BootstrapApplicationListener` class.

**Comment by mp911de** [comment]
@bhavanij gotcha! The issue is in `BootstrapApplicationListener`, not `OrderComparator`.

**Comment by mp911de** [comment]
This is by design. Spring Framework itself sorts instances anyway since 4.3.

**Comment by oialvarez** [comment]
@bhavanij Do you have the example with how you solved this? My requirement is using AppRole with credentials crypted using aes-256-cbc
